server/src/main/java/org/opensearch/gateway/GatewayAllocator.java

```diff
@@ -303,6 +313,59 @@ protected AsyncShardFetch.FetchResult<TransportNodesListGatewayStartedShards.Nod
         }
     }
 
+
+    class InternalPrimaryBatchShardAllocator extends PrimaryShardBatchAllocator {
+
+        @Override
+        @SuppressWarnings("unchecked")
+        protected AsyncShardFetch.FetchResult<TransportNodesListGatewayStartedShardsBatch.NodeGatewayStartedShardsBatch> fetchData(Set<ShardRouting> shardsEligibleForFetch,
```

Comment:
I am trying to understand, the method `fetchData(Set<ShardRouting> shardsEligibleForFetch, Set<ShardRouting> inEligibleShards, RoutingAllocation allocation)` is trying to fetch the Response for a single batch or across all the batches ? 

I am assuming it is former and the whole convulated logic of `shardId` -> `shardRouting` -> `batchID` -> `shardsBatch` -> `shardBatchState` is to not break the method definition, which we are overriding.  If this is true, then can we:
1. Rename `shardsEligibleForFetch` to `eligibleShardsInBatch` , `inEligibleShards` to `ineligibleShardsInBatch`
2.  Split this into 2 method, `fetchDataShardsBatch` and `fetchData` :
```
AsyncShardFetch.FetchResult<TransportNodesListGatewayStartedShardsBatch.NodeGatewayStartedShardsBatch> fetchDataShardsBatch (
    Set<ShardRouting> eligibleShardsInBatch,
    Set<ShardRouting> ineligibleShardsInBatch,
    RoutingAllocation allocation) {
  ...
  ShardsBatch shardsBatch = ...
  return fetchData(ShardsBatch, allocation);
}

AsyncShardFetch.FetchResult<TransportNodesListGatewayStartedShardsBatch.NodeGatewayStartedShardsBatch> fetchData (
    ShardsBatch shardsBatch,
    RoutingAllocation allocation) {
...
}
```

Reply:
1. Renaming part if fine.
2. On second point, that can be a good suggestion only if we have two such functions that cater to fetchData of both primaries and replicas, since both of the fetchData in original code is doing exact same things. Otherwise if follow the above approach as you suggested then dont you think it will overkill since it will lead to 4 different methods(2 each for replicas and primaries)that will be used at single place and have wont be having any reuse.
3. 
So extending on your though process, if we can do this then we can avoid some code duplication-
```
    @Override
        @SuppressWarnings("unchecked")
        protected AsyncShardFetch.FetchResult<TransportNodesListGatewayStartedBatchShards.NodeGatewayStartedShardsBatch> fetchData(
            Set<ShardRouting> shardsEligibleForFetch,
            Set<ShardRouting> inEligibleShards,
            RoutingAllocation allocation
        )
     ShardsBatch shardsbatch=fetchDataShardBatch(shardsEligibleForFetch, inEligibleShards)
     fetchDataForShardBatch(shardsbatch, shardsbatch.primary())

Same as above for replicas
```

And based on that we will implement two more methods `fetchDataShardBatch` to get batch of set of shards
and then a generic response from this method `fetchDataForShardBatch`, which will be later type casted by repected fetchData() call of primaries/replicas